### PR TITLE
Add a framework comparison doc

### LINKED ISSUE #638: Create a comparison guide with Streamlit / Gradio / Reflex

### PR DESCRIPTION
**Pull request by wwwillchen**
Closes #638

### CODE REVIEW

**Comment by wwwillchen** [comment]
@richard-to finally ready for review :)

Once you've taken a look, I'll share this with some Googlers and also the framework authors before publishing.

**Review by richard-to — APPROVED** [review]
My perspective is from someone who hasn't really used Gradio or Streamlit.

In reading this at a high level, I wasn't quite able to tell what was a pro/con between Mesop in some cases. So I think it reads somewhat objectively.

I think what's a bit confusing for me is that the document compares different aspects of Streamlit versus Gradio. I don't think that's a problem necessarily, but it could feel like cherry picking certain things. I don't think that's what's happening here. I think the doc is just pointing out the key differentiating features.

One thing I'd like to hear more about is performance  / scalability of Streamlit / Gradio / Mesop. I think that's an important use case.

Also what types of use cases are Streamlit / Gradio / Mesop typically used for.

Another thing I'd be curious of is extensibility. So if the core stuff is not sufficient, how much work is it to find/get the functionality I want into my app.

**Comment by wwwillchen** [comment]
Thanks for the review!

> My perspective is from someone who hasn't really used Gradio or Streamlit.
> 
> In reading this at a high level, I wasn't quite able to tell what was a pro/con between Mesop in some cases. So I think it reads somewhat objectively.
> 
> I think what's a bit confusing for me is that the document compares different aspects of Streamlit versus Gradio. I don't think that's a problem necessarily, but it could feel like cherry picking certain things. I don't think that's what's happening here. I think the doc is just pointing out the key differentiating features.

Yeah, I thought about this too and I think if we were to be comprehensive (and have the same criteria), it would be a much longer doc.

> 
> One thing I'd like to hear more about is performance / scalability of Streamlit / Gradio / Mesop. I think that's an important use case.

I think it's important but also requires benchmarking to be objective, speaking of which, we should do it ourselves :) Filed #706.

> 
> Also what types of use cases are Streamlit / Gradio / Mesop typically used for.

I mention this briefly - Streamlit historically has focused more on the data science use case, Gradio on ML use cases, but, tbh, everything seems to be converging as all three support LLM use cases.

> Another thing I'd be curious of is extensibility. So if the core stuff is not sufficient, how much work is it to find/get the functionality I want into my app.

Let me add this in - it's a big topic but I think the main differences are the UI frameworks that are well-supported.

File: docs/comparison.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+Mesop enables creating custom [web components](https://google.github.io/mesop/web_components/) based on open web standards, facilitating interoperability with components from different frameworks like Lit. Mesop web components are rendered in the same frame as the rest of the Mesop app which provides more flexibility but less isolation.
+
+Streamlit has a more established ecosystem of community-developed components, while Mesop's community and component ecosystem are still developing.
+
+## Gradio
+
+Gradio and Mesop both enable rapid ML/AI app development but with different approaches.
+
+Gradio has a strong focus on creating demos and interfaces for machine learning models and makes it easy to build a UI for a model.
```
Comment (abidlabs) [inline docs/comparison.md:41]:
```suggestion
Gradio has a strong focus on creating demos and interfaces for machine learning models and makes it easy to build a UI for a model. Gradio also offers a lower-level abstraction known as Blocks for more general web applications. 
```

**Comment by wwwillchen** [comment]
Thanks @abidlabs for the suggestions - I've incorporated them.

**Pull request merged**